### PR TITLE
Add JSON decode and encode for the lisk-codec - Close #5354

### LINKED ISSUE #5354: Add JSON decode and encode for the lisk-codec
### Description

For the JSON decode and encode,
`bytes` should be translated from/to base64 string.
`uint64/sint64` should be translated from/to number string.

interface should be same as encode, decode.

Add 
```
decodeJSON(schema: Schema, message: Buffer): JSONObject
encodeJSON(schema: Schema, message:JSONObject): Buffer
```

### Motivation

Although it will not be used in the protocol, it will be beneficial to have json decode and encode for debugging, using in HTTP plugin, or from the front end.

### Acceptance Criteria

- Test should be written using the protocol spec
- It should not be used in the framework

### PR DESCRIPTION
**Pull request by pablitovicente**
### What was the problem?

This PR resolves #5354

### How was it solved?

- by adding `encodeJSON`
- by adding `decodeJSON`
- by adding required type transformation maps
- by adding recursive methods to figure out the proper types

### How was it tested?

Full unit tests for both `encodeJSON` and `decodeJSON`

### CODE REVIEW

File: elements/lisk-codec/src/codec.ts [hunk 1]
```diff
@@ -113,6 +120,42 @@ export class Codec {
 		return (res as unknown) as T;
 	}
 
+	// For performance applications use decode() instead!
+	public decodeJSON = (schema: Schema, message: Buffer): GenericObject => {
```
Comment (shuse2) [inline elements/lisk-codec/src/codec.ts:124]:
```suggestion
	public decodeJSON<T>(schema: Schema, message: Buffer): T {
```

File: elements/lisk-codec/src/codec.ts [hunk 2]
```diff
@@ -113,6 +120,42 @@ export class Codec {
...
+			'toJSON',
+			decodedMessageCopy,
+			(schema as unknown) as SchemaProps,
+			[],
+		);
+		return decodedMessageCopy as GenericObject;
+	};
+
+	// For performance applications use encode() instead!
+	public encodeJSON = (
```
Comment (shuse2) [inline elements/lisk-codec/src/codec.ts:142]:
```
public encode(schema: Schema, message: object): Buffer
```

File: elements/lisk-codec/src/codec.ts (hunk 1 above)
Comment (pablitovicente) [reply elements/lisk-codec/src/codec.ts]: Had to make another adjustment so I committed from local repo. Done.

File: elements/lisk-codec/src/codec.ts (hunk 2 above)
Comment (pablitovicente) [reply elements/lisk-codec/src/codec.ts]: I think you meant the type for `message` which I changed. If it was not that please comment again :)

**Comment by pablitovicente** [comment]
> LGTM overall, just function signature should be the same as standard encode/decode.
> Also, it could be different issue but it would be useful to have conversion function also exposed.
> 
> ```
> public toJSON<T = object>(schema: Schema, message: object): T {}
> public toObject<T = object>(schema: Schema, message: object): T {}
> ```

Will split in different issue and start it in parallel to this PR

File: elements/lisk-codec/src/codec.ts [hunk 3]
```diff
@@ -113,6 +120,39 @@ export class Codec {
 		return (res as unknown) as T;
 	}
 
+	// For performance applications use decode() instead!
```
Comment (nazarhussain) [inline elements/lisk-codec/src/codec.ts:123]: I don't get this comment. `decodeJSON` and `decode` have very different use cases. Can't replace each other for performance reasons.

File: elements/lisk-codec/src/json_wrapper.ts [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+	GenericObject,
+	SchemaPair,
+	SchemaProps,
+	SchemaScalarItem,
+} from './types';
+
+export interface IteratableGenericObject extends GenericObject {
+	// eslint-disable-next-line @typescript-eslint/no-explicit-any
+	[Symbol.iterator](): Iterator<{ key: string; value: any }>;
+}
```
Comment (nazarhussain) [inline elements/lisk-codec/src/json_wrapper.ts:27]: Should be moved to `types`

File: elements/lisk-codec/src/codec.ts [hunk 5]
```diff
@@ -113,6 +120,39 @@ export class Codec {
...
+			schema,
+			message,
+		);
+		const decodedMessageCopy = objectUtils.cloneDeep(decodedMessage);
+		decodedMessageCopy[Symbol.iterator] = iterator;
+
+		recursiveTypeCast(
+			'toJSON',
+			decodedMessageCopy,
+			(schema as unknown) as SchemaProps,
```
Comment (nazarhussain) [inline elements/lisk-codec/src/codec.ts:135]: Why not pass `scehma` as it is, and treat it as you want inside the function?

File: elements/lisk-codec/src/json_wrapper.ts [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+		done: boolean;
+		value: { value: IteratableGenericObject; key: string };
+	};
+} {
+	let index = 0;
+	const properties = Object.keys(this);
+	let Done = false;
+	return {
+		// eslint-disable-next-line @typescript-eslint/no-explicit-any
+		next: (): { done: boolean; value: { value: any; key: string } } => {
```
Comment (nazarhussain) [inline elements/lisk-codec/src/json_wrapper.ts:91]: The return type for `value` in both places should match.

File: elements/lisk-codec/src/codec.ts [hunk 7]
```diff
@@ -113,6 +120,39 @@ export class Codec {
...
+
+	// For performance applications use encode() instead!
+	public encodeJSON(schema: Schema, message: object): Buffer {
+		const messageCopy = objectUtils.cloneDeep(message);
+		(messageCopy as IteratableGenericObject)[Symbol.iterator] = iterator;
+
+		recursiveTypeCast(
+			'fromJSON',
+			messageCopy as IteratableGenericObject,
+			(schema as unknown) as SchemaProps,
```
Comment (nazarhussain) [inline elements/lisk-codec/src/codec.ts:149]: Can we use compiled schema here? If it's efficient for `encode` we may have its benefit here as well.

File: elements/lisk-codec/src/json_wrapper.ts [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+						throw new Error(
+							`Invalid schema property found. Path: ${dataPath.join(',')}`,
+						);
+					}
+					if (mode === 'toJSON') {
+						// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
+						(object[key] as any)[i] = _liskMessageValueToJSONValue[
+							(schemaProp.items as SchemaScalarItem).dataType
+						](value[i]);
+					} else {
```
Comment (nazarhussain) [inline elements/lisk-codec/src/json_wrapper.ts:146]: This if/else inside loop and recursive calls does not look good and would be less performant. It would be nicer if we split the implementation into two functions.

File: elements/lisk-codec/src/json_wrapper.ts [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+};
+
+export const recursiveTypeCast = (
+	mode: 'toJSON' | 'fromJSON',
+	object: IteratableGenericObject,
+	schema: SchemaProps,
+	dataPath: string[],
+): void => {
+	for (const { key, value } of object) {
+		if (isObject(value)) {
```
Comment (nazarhussain) [inline elements/lisk-codec/src/json_wrapper.ts:111]: Instead of this native check can' we use schema defined `type: object`

File: elements/lisk-codec/src/json_wrapper.ts [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+
+			(value as IteratableGenericObject)[Symbol.iterator] = iterator;
+			recursiveTypeCast(mode, value, schema, dataPath);
+			dataPath.pop();
+
+			delete (value as IteratableGenericObject)[
+				(Symbol.iterator as unknown) as string
+			];
+		} else if (Array.isArray(value)) {
+			dataPath.push(key);
```
Comment (nazarhussain) [inline elements/lisk-codec/src/json_wrapper.ts:122]: Same here.

File: elements/lisk-codec/src/codec.ts (hunk 3 above)
Comment (pablitovicente) [reply elements/lisk-codec/src/codec.ts:123]: It's just in case someone thinks it can just use it directly. Imagine someone is working in a custom plugin and thinks "hey I can just use my object directly with this methods decodeJSON/encodeJSON" that might be or not what they want. So it's there as a warning to call the attention of the user.

File: elements/lisk-codec/src/codec.ts (hunk 7 above)
Comment (pablitovicente) [reply elements/lisk-codec/src/codec.ts:149]: Changing to compiled schema will mean a rewrite of this whole PR and was discussed with Shu for now not to do it. We can refactor in the future to use compiled schema.

File: elements/lisk-codec/src/codec.ts (hunk 5 above)
Comment (pablitovicente) [reply elements/lisk-codec/src/codec.ts:135]: Sorry @nazarhussain I do not understand this question could you rephrase/expand please?

File: elements/lisk-codec/src/json_wrapper.ts (hunk 9 above)
Comment (pablitovicente) [reply elements/lisk-codec/src/json_wrapper.ts:111]: This is a check to go into recursion if value is an object literal; I think the TS type will not work for this?

File: elements/lisk-codec/src/json_wrapper.ts (hunk 10 above)
Comment (pablitovicente) [reply elements/lisk-codec/src/json_wrapper.ts:122]: Same question as for the one there.

File: elements/lisk-codec/src/json_wrapper.ts (hunk 8 above)
Comment (pablitovicente) [reply elements/lisk-codec/src/json_wrapper.ts:146]: I get the point but if we split we end up with two functions that are exactly the same but in 1 line. Will check if possible to switch the map dynamically based on the parameter with little changes.

File: elements/lisk-codec/src/codec.ts (hunk 3 above)
Comment (nazarhussain) [reply elements/lisk-codec/src/codec.ts:123]: I think it depends upon the use case if one needs JSON, it will use it, others don't.

File: elements/lisk-codec/src/json_wrapper.ts (hunk 9 above)
Comment (pablitovicente) [reply elements/lisk-codec/src/json_wrapper.ts:111]: OK got clarification about this. I prefer to keep the check by observing the object itself.

File: elements/lisk-codec/src/codec.ts (hunk 5 above)
Comment (nazarhussain) [reply elements/lisk-codec/src/codec.ts:135]: You are typecasting `schema` to `SchemaProps`, looks like both are same otherwise would not work.  So why not pass the schema without `as SchemaProps`

File: elements/lisk-codec/src/json_wrapper.ts (hunk 8 above)
Comment (nazarhussain) [reply elements/lisk-codec/src/json_wrapper.ts:146]: But its inside loop and recursion. You yourself are emphasizing on the fact the even slightest performance really matters for `lisk-codec` then why not here.

File: elements/lisk-codec/src/json_wrapper.ts (hunk 9 above)
Comment (nazarhussain) [reply elements/lisk-codec/src/json_wrapper.ts:111]: No not TS type. If its an object, then relevant schema definition must have `type: object` for it. So why not use that?

File: elements/lisk-codec/src/json_wrapper.ts (hunk 8 above)
Comment (pablitovicente) [reply elements/lisk-codec/src/json_wrapper.ts:146]: @nazarhussain I emphasise it for `encode()` and `decode()` this methods are extras.

File: elements/lisk-codec/src/json_wrapper.ts (hunk 4 above)
Comment (pablitovicente) [reply elements/lisk-codec/src/json_wrapper.ts:27]: Moved.

File: elements/lisk-codec/src/json_wrapper.ts (hunk 6 above)
Comment (pablitovicente) [reply elements/lisk-codec/src/json_wrapper.ts]: Fixed

File: elements/lisk-codec/src/codec.ts (hunk 3 above)
Comment (pablitovicente) [reply elements/lisk-codec/src/codec.ts:120]: I think the comment does not harm. When documentation is written for lisk-codec maybe this comment can be removed. I think it doesn't hurt to leave it for now.

File: elements/lisk-codec/src/json_wrapper.ts (hunk 8 above)
Comment (pablitovicente) [reply elements/lisk-codec/src/json_wrapper.ts:146]: Changed to a map to make it constant time but still sorting out the type for the new map.

File: elements/lisk-codec/src/json_wrapper.ts [hunk 11]
```diff
@@ -15,15 +15,16 @@
...
 	BaseTypes,
-	GenericObject,
+	IteratableGenericObject,
 	SchemaPair,
 	SchemaProps,
 	SchemaScalarItem,
 } from './types';
 
-export interface IteratableGenericObject extends GenericObject {
+interface iteratorReturnValue {
```
Comment (nazarhussain) [inline elements/lisk-codec/src/json_wrapper.ts:24]: Types always capitalize. 

```suggestion
interface IteratorReturnValue {
```

File: elements/lisk-codec/src/types.ts [hunk 12]
```diff
@@ -71,3 +71,8 @@ export interface CompiledSchemas {
 export interface Validator {
 	addMetaSchema: (schema: object, key?: string) => {};
 }
+
+export interface IteratableGenericObject extends GenericObject {
+	// eslint-disable-next-line @typescript-eslint/no-explicit-any
+	[Symbol.iterator](): Iterator<{ key: string; value: any }>;
```
Comment (nazarhussain) [inline elements/lisk-codec/src/types.ts:77]:
```suggestion
	[Symbol.iterator](): Iterator<IteratorReturnValue>;
```

File: elements/lisk-codec/src/json_wrapper.ts (hunk 8 above)
Comment (pablitovicente) [reply elements/lisk-codec/src/json_wrapper.ts]: Done. Resolving.

File: elements/lisk-codec/src/codec.ts (hunk 5 above)
Comment (pablitovicente) [reply elements/lisk-codec/src/codec.ts:132]: Let's leave as is. Maybe let's discuss complete type refactor for lisk-codec if deemed necessary as this abstract structures structures might need better type abastractions.

File: elements/lisk-codec/src/types.ts (hunk 12 above)
Comment (pablitovicente) [reply elements/lisk-codec/src/types.ts]: Had to do it locally as I had to move the IteratorReturnValue to the types file.
